### PR TITLE
Add validation of smt parse trees to construct smt_responset instances.

### PR DESCRIPTION
**Pull request by thomasspriggs**
This PR adds validation of smt parse trees to construct smt_responset instances.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/solvers/smt2_incremental/smt_response_validation.cpp [hunk 1]
```diff
@@ -37,5 +37,14 @@ template class response_or_errort<smt_responset>;
 
 response_or_errort<smt_responset> validate_smt_response(const irept &parse_tree)
 {
+  if(parse_tree.id() == "sat")
```
Comment (NlightNFotis) [inline src/solvers/smt2_incremental/smt_response_validation.cpp:309]: ⛏️ Why not `switch (parse_tree.id()) { case "sat": }`. Might make it a bit more readable?

**Review by NlightNFotis — COMMENTED** [review]
LGTM overall 👍🏻

File: src/solvers/smt2_incremental/smt_response_validation.cpp (hunk 1 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt_response_validation.cpp:309]: Because you can't switch on a string in C++.

Comment (NlightNFotis) [reply src/solvers/smt2_incremental/smt_response_validation.cpp:309]: 😱

**Review by NlightNFotis — APPROVED** [review]

File: src/solvers/smt2_incremental/smt_response_validation.cpp [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+  // unexpected in the parse tree is now considered to be an invalid response.
+  if(parse_tree.get_sub().size() == 1)
+  {
+    return {response_or_errort<smt_responset>{
+      "Error response is missing the error message."}};
+  }
+  if(parse_tree.get_sub().size() > 2)
+  {
+    return {response_or_errort<smt_responset>{
+      "Error response has multiple error messages."}};
```
Comment (TGWDB) [inline src/solvers/smt2_incremental/smt_response_validation.cpp]: Can we do something nicer here? Maybe gather all the error messages into a multiline string?

File: src/solvers/smt2_incremental/smt_response_validation.cpp (hunk 1 above)
Comment (TGWDB) [reply src/solvers/smt2_incremental/smt_response_validation.cpp:309]: But you can switch on a/the first character... so if you want a big nested tree of case statements... ;)

**Review by TGWDB — APPROVED** [review]

File: src/solvers/smt2_incremental/smt_response_validation.cpp (hunk 2 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt_response_validation.cpp]: I have done something nicer here. The original parse tree is now included in this error message, as instances where the messages are sub trees could also fall into this case.

**Pull request merged**